src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,151 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    public MyLinkedList() {
+
+    }
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            Node<T> node = new Node<>(null, value, null);
+            head = node;
+            tail = node;
+            size++;
+            return;
+        }
+
+        Node<T> prev = tail;
+        Node<T> node = new Node<>(prev, value, null);
+        tail = node;
+        prev.next = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size && index >= 0) {
+            add(value);
+            return;
+        }
+        Node<T> oldNode = getNodeByIndex(index);
+        Node<T> newNode = new Node<>(oldNode.prev, value, oldNode);
+
+        if (index == 0) {
+            head = newNode;
+        } else {
+            oldNode.prev.next = newNode;
+        }
+        oldNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNodeByIndex(index);
+        T oldNodeItem = node.item;
+        node.item = value;
+        return oldNodeItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeToRemove = getNodeByIndex(index);
+        unlink(nodeToRemove);
+        return nodeToRemove.item;
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> i = head; i != null; i = i.next) {
+            if (object != null && object.equals(i.item)
+                    || object == i.item) {
+                unlink(i);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        indexChecking(index);
+        Node<T> node;
+        if (index < (size >> 1)) {
+            node = head;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+            return node;
+        }
+
+        node = tail;
+        for (int i = size - 1; i > index; i--) {
+            node = node.prev;
+        }
+        return node;
```

Comment:
let's make it more readable :)
```suggestion
        indexChecking(index);
        Node<T> x;
        if (index < size / 2) {
            x = head;
            for (int i = 0; i < index; i++) {
                x = x.next;
            }
        } else {
            x = tail;
            for (int i = size - 1; i > index; i--) {
                x = x.prev;
            }
        }
        return x;
```